app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt

```diff
@@ -0,0 +1,418 @@
+                        },
+                        trailingIcon = {
+                            IconButton(onClick = { filterText = TextFieldValue() }) {
+                                Icon(imageVector = Icons.Filled.Cancel,
+                                     tint = MaterialTheme.colors.primary,
+                                     contentDescription = null)
+                            }
+                        },
+                    )
+                    Row(
+                        modifier = Modifier.fillMaxWidth().padding(horizontal = 8.dp),
+                    ) {
+                        @Composable
+                        fun <T : PuzzleDirectorySort> ratingButton(type: KClass<T>, name: String) {
+                            val sorting = type.safeCast(sortField)
+                            Spacer(modifier = Modifier.weight(1f))
+                            TextButton(onClick = {
+                                sortField = sorting?.reversed ?: type.createInstance()
+                            }) {
+                                Text(
+                                    text = name,
+                                    color = if(sorting == null) MaterialTheme.colors.secondary
+                                            else MaterialTheme.colors.primary,
+                                    fontSize = 12.sp)
+                                Icon(
+                                    imageVector = if(sorting?.asc == false) Icons.Filled.ExpandMore
+                                                  else Icons.Filled.ExpandLess,
+                                    tint = if(sorting == null) MaterialTheme.colors.secondary
+                                           else MaterialTheme.colors.primary,
+                                    modifier = Modifier
+                                       .align(Alignment.CenterVertically),
+                                    contentDescription = null)
+                            }
+                        }
+                        Text(
+                            text = "Sort:",
+                            color = MaterialTheme.colors.onBackground,
+                            fontSize = 14.sp,
+                            fontWeight = FontWeight.Bold,
+                            modifier = Modifier
+                                .align(Alignment.CenterVertically)
+                        )
+                        ratingButton(name = "Name", type = NameSort::class)
+                        ratingButton(name = "Rating", type = RatingSort::class)
+                        ratingButton(name = "Count", type = CountSort::class)
+                        ratingButton(name = "Views", type = ViewsSort::class)
+                        ratingButton(name = "Rank", type = RankSort::class)
+                    }
+
+                    if (filterText.text.isEmpty()) {
+                        LazyRow(modifier = Modifier.fillMaxWidth()) {
```

Comment:
It's horizontal, because it displays all the visited puzzle sets, but you'll have to horizontally scroll to see ones further back in history - try visiting more than three to see an example. It's lazy so that that's cheap. I don't think it's a problem, the horizontal scroll isn't even accessible unless you know its there, and as I said is free, but up to you if you want it

Reply:
There might be something to consider in dropping the gap when there's less then three visited, I agree that's kinda unnecessary and garish